automatic/sourcetree/update_helper.ps1

```diff
@@ -1,54 +1,42 @@
-function Get-PackageName() {
-  param(
-    [string]$a
-  )
-
-  switch -w ( $a ) {
-	  'sourcetree' {
-		  $PackageUrl = "https://www.sourcetreeapp.com/enterprise"
-	  }
-  }
-
-  return $PackageUrl
-}
 
 function Get-JavaSiteUpdates {
   param(
     [string]$Package,
     [string]$Title,
-    [string]$Wait = 4
+    [int]$revision = 20 # This is updatable outside the function should the number need to go higher
   )
-
-  $regex = '([\d]{0,2}[\.][\d]{0,2}[\.][\d]{0,2}[\.][\d]{0,5})'
-  $url = Get-PackageName $Package
-  $ie = New-Object -comobject InternetExplorer.Application
-  $ie.Navigate2($url) 
-  $ie.Visible = $false
-
-  while($ie.ReadyState -ne $Wait) {
-    Start-Sleep -Seconds 20
-  }
-
-  foreach ( $_ in $ie.Document.getElementsByTagName("a") ) {
-    $url = $_.href;
-    if ( $url -match $regex) {
-      $yes = $url | Select-Object -last 1
-      $version = $Matches[0]
-      break;
+  
+# We are using this url to test for new versions since it is updated with the latest version as soon as it is available
+$releaseNotesURL = "https://product-downloads.atlassian.com/software/sourcetree/windows/ga/ReleaseNotes_<version>.html"
+# Next we will read the last good version from the nuspec file
+$PreVersion = ([xml](Get-Content "$PSScriptRoot\sourcetree.nuspec")).package.metadata.version
+#  Now to split the PreVersion for the loops
+[int]$z,[int]$y,[int]$x = $PreVersion -split('\.')
+for($a=$z; $a -lt $revision;$a++){
+  for($b=$y; $b -lt $revision;$b++){
+    for($c=$x; $c -lt $revision;$c++){
+      $testing = $releaseNotesURL -replace( "<version>" ,  -join($a, ".", $b, ".", $c ) )
+      try {
+        $url = Invoke-WebRequest -UseBasicParsing -Uri $testing | Select-Object StatusCode
+      }
+      catch {
+        if (!([string]::IsNullOrEmpty( $url )) ) {
+          # This will get the number that fails by subtracting 1 from c we will get the latest version
+          $version = -join($a, ".", $b, ".", ($c -1) )
+          $version | Write-Warning
+          # This is an emergency stop of all for loops
+          $revision = 0
+        }
+      }
```

Comment:
Honestly, I'm not fond of this approach at all.
In the worst-case scenario, this could result in making ~8000 (maybe more) requests upstream just for a single package, and it is not something I could, with good conscience, accept into this repository.

A different approach would need to be found for parsing the version.

Reply:
Please, more than 8k requests is a bit far fetched of an exaggeration. The most it did in testing is 100 requests, but that was an extreme test with NO starting point. 
I'll look for a less resource usage intensive process to find this information. 

I have a question does the version number have to come from the source of the package or can it be gotten from a third party site (ex: FileHippo)?

Reply:
I was going to say that we could relatively simply scrape [this page](https://www.sourcetreeapp.com/enterprise) or [this page](https://www.sourcetreeapp.com/download-archives), but do note that neither of those result in the "latest" version according to the code above (3.4.11, seemingly).

That said, if clicking the `Download for Windows` button on the homepage (or checking [FileHippo](https://filehippo.com/download_sourcetree_win/), for instance) also results in that slightly lower version, I guess I'm unsure if the currently-visible 3.4.11 release notes are actually for a prerelease?

Reply:
Does the software auto-update? If so, can we look to use the endpoint that it queries?

> In the worst-case scenario, this could result in making ~8000 (maybe more) requests upstream 
> Please, more than 8k requests is a bit far fetched of an exaggeration

It was mentioned that this was a 'worst-case scenario'. Let's take Kim's guidance and experience in this rather than shooting him down. Looking for a less resource intensive way to do it is appreciated.

Reply:
I'll start off by saying I'm not a user of this package or its software. If someone knows how the software phones home. I would be interested to learn more.  

I'll look into another approach for getting the latest version number in a less resource intensive process and get back to you all.